README.md
ABP Core Repository
=====================================



ABP is a cryptocurrency and platfrom that will serve to decentralize the media and entertainement industry.  It uses a custom Proof of Stake protocol for securing its network and uses an innovative variable seesaw reward mechanism that dynamically balances 70% of its block reward size to masternodes, 10% to staking nodes 15% to budget proposals, and 5% to developments. The goal of ABP is to achieve a decentralization of motion picture funding, and create an ecosystem governed by Initial Video Offerings and masternode governance.

### Features

- Anonymous transactions with Zerocoin Protocol.
- Fast transactions featuring guaranteed zero confirmation transactions, we call it _FastTX_.
- Decentralized blockchain voting using Masternodes for project proposals, which are funded by the governance. The blockchain will distribute governance funds based on successful proposals submitted by the community and voted on by the masternode holders.

More information at [abpnetwork.io](http://www.abpnetwork.io).

### Coin Specs
<table>
<tr><td>Block Time</td><td>60 Seconds</td></tr>
<tr><td>Difficulty Retargeting</td><td>Every Block</td></tr>
<tr><td>Max Coin Supply</td><td>21,000,000</td></tr>
<tr><td>Premine</td><td>1,050,000 ABP</td></tr>
<tr><td>Network</td><td>PoS 3.0</td></tr>
<tr><td>Privacy Technology</td><td>Zerocoin Protocol</td></tr>
<tr><td>Fast Send</td><td>FastTX</td></tr>
<tr><td>Minimum Stake Age</td><td>3h</td></tr>
<tr><td>Transaction Fee</td><td>0.001 ABP</td></tr>
<tr><td>Fast Send</td><td>FastTX</td></tr>
<tr><td>Confirmations</td><td>6</td></tr>
<tr><td>Accumulator Encryption</td><td>RSA-2048</td></tr>
<tr><td>MN Collateral</td><td>1000 ABP</td></tr>
<tr><td>Block Rewards</td><td>2-10 ABP (Wave Sequence)</td></tr>
</table>

### Reward Distribution

<table>
<th colspan=4>Genesis Block</th>
<tr><th>Block Height</th><th>Reward Amount</th><th>Notes</th></tr>
<tr><td>1</td><td>1,050,000 ABP</td></tr>
</table>

### PoS, Governance, Developments and Masternodes Reward Schedule

<table>
<th>Phase</th><th>Block Height</th><th>Masternodes (70%)</th><th>PoS (10%)</th><th>Governance (15%)</th><th>Developments (5%)</th>
<tr><td>Phase 1</td><td>2-10080</td><td>1.4 ABP</td><td>0.2 ABP</td><td>0.3 ABP</td><td>0.1 ABP</td></tr>
<tr><td>Phase 2</td><td>10081-20160</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 3</td><td>20161-30240</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 4</td><td>30241-40320</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 5</td><td>40321-50400</td><td>7 ABP</td><td>1 ABP</td><td>1.5 ABP</td><td>0.5 ABP</td></tr>
<tr><td>Phase 6</td><td>50401-60480</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 7</td><td>60481-70560</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 8</td><td>70561-80640</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 9</td><td>80641-90720</td><td>1.4 ABP</td><td>0.2 ABP</td><td>0.3 ABP</td><td>0.1 ABP</td></tr>
<tr><td>Phase 10</td><td>90721-100800</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 11</td><td>100801-110880</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 12</td><td>110881-120960</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 13</td><td>120961-131040</td><td>7 ABP</td><td>1 ABP</td><td>1.5 ABP</td><td>0.5 ABP</td></tr>
<tr><td>Phase 14</td><td>131041-141120</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 15</td><td>141121-151200</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 16</td><td>151201-161280</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 17</td><td>161281-171360</td><td>1.4 ABP</td><td>0.2 ABP</td><td>0.3 ABP</td><td>0.1 ABP</td></tr>
<tr><td>Phase 18</td><td>171361-181440</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 19</td><td>181441-191520</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 20</td><td>191521-201600</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 21</td><td>201601-211680</td><td>7 ABP</td><td>1 ABP</td><td>1.5 ABP</td><td>0.5 ABP</td></tr>
<tr><td>Phase 22</td><td>211681-221760</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 23</td><td>221761-231840</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 24</td><td>231841-241920</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 25</td><td>241921-252000</td><td>1.4 ABP</td><td>0.2 ABP</td><td>0.3 ABP</td><td>0.1 ABP</td></tr>
<tr><td>Phase 26</td><td>252001-262080</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 27</td><td>262081-272160</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 28</td><td>272161-282240</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 30</td><td>282241-292320</td><td>7 ABP</td><td>1 ABP</td><td>1.5 ABP</td><td>0.5 ABP</td></tr>
<tr><td>Phase 31</td><td>292321-302400</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 32</td><td>302401-312480</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 33</td><td>312481-322560</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 34</td><td>322561-332640</td><td>1.4 ABP</td><td>0.2 ABP</td><td>0.3 ABP</td><td>0.1 ABP</td></tr>
<tr><td>Phase 35</td><td>332641-342720</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 36</td><td>342721-352800</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 37</td><td>352801-362880</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 38</td><td>362881-372960</td><td>7 ABP</td><td>1 ABP</td><td>1.5 ABP</td><td>0.5 ABP</td></tr>
<tr><td>Phase 39</td><td>372961-383040</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 40</td><td>383041-393120</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 41</td><td>393121-403200</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 42</td><td>403201-413280</td><td>1.4 ABP</td><td>0.2 ABP</td><td>0.3 ABP</td><td>0.1 ABP</td></tr>
<tr><td>Phase 43</td><td>413281-423360</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 44</td><td>423361-433440</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 45</td><td>433441-443520</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 46</td><td>443521-453600</td><td>7 ABP</td><td>1 ABP</td><td>1.5 ABP</td><td>0.5 ABP</td></tr>
<tr><td>Phase 47</td><td>453601-463680</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 48</td><td>463681-473760</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 49</td><td>473761-483840</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 50</td><td>483841-493920</td><td>1.4 ABP</td><td>0.2 ABP</td><td>0.3 ABP</td><td>0.1 ABP</td></tr>
<tr><td>Phase 51</td><td>493921-504000</td><td>2.8 ABP</td><td>0.4 ABP</td><td>0.6 ABP</td><td>0.2 ABP</td></tr>
<tr><td>Phase 52</td><td>504001-514080</td><td>4.2 ABP</td><td>0.6 ABP</td><td>0.9 ABP</td><td>0.3 ABP</td></tr>
<tr><td>Phase 53</td><td>514081-524160</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
<tr><td>Phase 54</td><td>524161-Max Supply</td><td>5.6 ABP</td><td>0.8 ABP</td><td>1.2 ABP</td><td>0.4 ABP</td></tr>
</table>


doc/release-notes.md
(note: this is a temporary file, to be added-to by anybody, and moved to release-notes at release time)

ABP Core version *version* is now available from:  <https://github.com/abp-project/abp/releases>

This is a new major version release, including various bug fixes and performance improvements, as well as updated translations.

Please report bugs using the issue tracker at github: <https://github.com/abp-project/abp/issues>


Mandatory Update
==============


How to Upgrade
==============

If you are running an older version, shut it down. Wait until it has completely shut down (which might take a few minutes for older versions), then run the installer (on Windows) or just copy over /Applications/ABP-Qt (on Mac) or abpd/abp-qt (on Linux).


Compatibility
==============

ABP Core is extensively tested on multiple operating systems using the Linux kernel, macOS 10.8+, and Windows 7 and later.

Microsoft ended support for Windows XP on [April 8th, 2014](https://www.microsoft.com/en-us/WindowsForBusiness/end-of-xp-support), No attempt is made to prevent installing or running the software on Windows XP, you can still do so at your own risk but be aware that there are known instabilities and issues. Please do not report issues about Windows XP to the issue tracker.

ABP Core should also work on most other Unix-like systems but is not frequently tested on them.

### :exclamation::exclamation::exclamation: MacOS 10.13 High Sierra :exclamation::exclamation::exclamation:

**Currently there are issues with the 3.x gitian release on MacOS version 10.13 (High Sierra), no reports of issues on older versions of MacOS.**

 
Notable Changes
==============


*version* Change log
==============

Detailed release notes follow. This overview includes changes that affect behavior, not code moves, refactors and string updates. For convenience in locating the code changes and accompanying discussion, both the pull request and git merge commit are mentioned.

### Core Features

### Build System
 
### P2P Protocol and Network Code

### GUI
 
### RPC/REST

### Wallet
 
### Miscellaneous
 
 
## Credits

Thanks to everyone who directly contributed to this release:


As well as everyone that helped translating on [Transifex](https://www.transifex.com/projects/p/abp-project-translations/), the QA team during Testing and the Node hosts supporting our Testnet.


src/bip38.h
// Copyright (c) 2017 The ABP developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#ifndef BITCOIN_BIP38_H
#define BITCOIN_BIP38_H

#include "pubkey.h"
#include "uint256.h"

#include <string>


/** 39 bytes - 78 characters
 * 1) Prefix - 2 bytes - 4 chars - strKey[0..3]
 * 2) Flagbyte - 1 byte - 2 chars - strKey[4..5]
 * 3) addresshash - 4 bytes - 8 chars - strKey[6..13]
 * 4) Owner Entropy - 8 bytes - 16 chars - strKey[14..29]
 * 5) Encrypted Part 1 - 8 bytes - 16 chars - strKey[30..45]
 * 6) Encrypted Part 2 - 16 bytes - 32 chars - strKey[46..77]
 */

void DecryptAES(uint256 encryptedIn, uint256 decryptionKey, uint256& output);

void ComputePreFactor(std::string strPassphrase, std::string strSalt, uint256& prefactor);

void ComputePassfactor(std::string ownersalt, uint256 prefactor, uint256& passfactor);

bool ComputePasspoint(uint256 passfactor, CPubKey& passpoint);

void ComputeSeedBPass(CPubKey passpoint, std::string strAddressHash, std::string strOwnerSalt, uint512& seedBPass);

void ComputeFactorB(uint256 seedB, uint256& factorB);

std::string BIP38_Encrypt(std::string strAddress, std::string strPassphrase, uint256 privKey, bool fCompressed);
bool BIP38_Decrypt(std::string strPassphrase, std::string strEncryptedKey, uint256& privKey, bool& fCompressed);

std::string AddressToBip38Hash(std::string address);

#endif // BIP38_H
